apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/AbstractSpanInstrumentation.java

```diff
@@ -178,8 +178,9 @@ public CaptureExceptionInstrumentation() {
         @VisibleForAdvice
         @Advice.OnMethodExit(suppress = Throwable.class)
         public static void captureException(@Advice.FieldValue(value = "span", typing = Assigner.Typing.DYNAMIC) TraceContextHolder<?> context,
-                                            @Advice.Argument(0) Throwable t) {
-            context.captureException(t);
+                                            @Advice.Argument(0) Throwable t,
+                                            @Advice.Return(readOnly = false) String errorId) {
```

Comment:
This probably won't work when a user upgrades the agent and not the API (a very valid scenario)- the agent will instrument the `captureException` API that returns void and the instrumentation will expect a String being returned.

What you need to do is change the constructor of this class, so that the method matcher is more specific on the return type, probably something like: `named("captureException").and(takesArguments(Throwable.class)).and(returns(TypeDescription.VOID))` and add another `CaptureExceptionWithReturnIdInstrumentation` class that uses a method matcher like: `named("captureException").and(takesArguments(Throwable.class)).and(returns(TypeDescription.STRING))` and does what you did here. 

Eventually, you must test your branch's agent with an older API jar, and you should test your branch's API jar with an older agent. We don't have automatic tests for that, but should be easy to do manually.

Reply:
added new method captureExceptionAndReturnErrorId